### ISSUE TITLE
Proposal: option for consistent-return to not treat return in generator functions as error

### ISSUE THREAD

**Issue by ek5000** [eslint/eslint#11371]
**What rule do you want to change?**

consistent-return

**Does this change cause the rule to produce more or fewer warnings?**

fewer

**How will the change be implemented? (New option, new default behavior, etc.)?**

New option

**Please provide some example code that this change will affect:**

```js
function* exampleGenerator(option) {
    if (option === false) {
        return [];
    }

    yield* [1, 2, 3];
}
```

**What does the rule currently do for this code?**

Currently, this will produce an error, as there isn't an error for all branches of code. However, returning in a generator is a valid way to exhaust a generator early.

**What will the rule do after it's changed?**

With an option to the rule (something like `"ignoreGenerators": true`), the rule could ignore generator functions. I think an argument could be made that for generator functions, yield statements could be treated the same as return statements, at least for the purposes of this rule, but that might be a larger change.

**Are you willing to submit a pull request to implement this change?**

Yes.

**Comment by aladdin-add** [comment]
Thanks for creating the issue! sounds good to me. hopefully I'd like to propose the option can be enabled by default.(it might be a breaking change).

I will champion this.

**Comment by not-an-aardvark** [comment]
Personally, I think it would be better to update the default behavior here rather than adding a new option.

**Comment by kaicataldo** [comment]
@eslint/eslint-team Looks like we need one more :+1: to accept this, or it's probably time to close it.

**Comment by kaicataldo** [comment]
Whoops, sorry! Didn't mean to close it just now 😅

**Comment by mdjermanovic** [comment]
PR #12356

**Comment by mdjermanovic** [comment]
Can we clarify why should this rule ignore generators? 

If the value in `{ done: true }` object is a part of the API for the particular generator, it makes sense to warn when there's an inconsistency like in the original example. 

If it isn't, `consistent-return` already allows this:

```js
/* eslint consistent-return: error */

function* exampleGenerator(option) {
    if (option === false) {
        return;
    }

    yield* [1, 2, 3];
}

// no errors
```

**Comment by btmills** [comment]
I'm with @mdjermanovic here in thinking that the rule is working correctly as-is.

```js
function* exampleGenerator(option) {
    if (option === false) {
        return [];
    }

    yield* [1, 2, 3];
}
```

In the given example, sometimes the function returns an empty array, and sometimes it implicitly returns `undefined` at the end of iteration, which is exactly what `consistent-return` is supposed to prevent. To make that distinction more explicit, the type annotation for `exampleGenerator` [in TypeScript](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-6.html#stricter-generators) would be `Generator<number, Array<never> | undefined>`, where it iterates over `number`s but then `return`s either an empty array or `undefined`.

**Comment by nzakas** [comment]
Agreed, I don't think we need to make a change here. Thanks everyone.
